multicast/src/test/kotlin/com/dropbox/flow/multicast/ChannelManagerTest.kt

```diff
@@ -15,64 +15,154 @@
  */
 package com.dropbox.flow.multicast
 
-import com.dropbox.flow.multicast.ChannelManager.Message.AddChannel
-import com.dropbox.flow.multicast.ChannelManager.Message.DispatchValue
-import com.dropbox.flow.multicast.ChannelManager.Message.RemoveChannel
-import kotlinx.coroutines.CompletableDeferred
+import com.dropbox.flow.multicast.ChannelManager.Message.Dispatch
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.FlowPreview
 import kotlinx.coroutines.async
 import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.consumeAsFlow
-import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onEach
 import kotlinx.coroutines.flow.take
 import kotlinx.coroutines.flow.toList
-import kotlinx.coroutines.suspendCancellableCoroutine
 import kotlinx.coroutines.test.TestCoroutineScope
 import kotlinx.coroutines.test.runBlockingTest
 import org.assertj.core.api.Assertions.assertThat
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.junit.runners.JUnit4
+import java.lang.Exception
 
 @FlowPreview
 @ExperimentalCoroutinesApi
 @RunWith(JUnit4::class)
 class ChannelManagerTest {
     private val scope = TestCoroutineScope()
-    private val manager = ChannelManager<String>(
-            scope,
-            0,
-            onEach = {}
-    ) {
-        SharedFlowProducer(
-                scope, src =
-        flow {
-            suspendCancellableCoroutine<String> {
-                // never end
+    private val upstream: Channel<String> = Channel(Channel.UNLIMITED)
+    private val manager = ChannelManager(
+        scope,
+        0,
+        onEach = {},
+        upstream = upstream.consumeAsFlow()
+    )
+
+    @Test
+    fun `Given one downstream WHEN two values come in on the upstream THEN two values are consumed`() =
+        scope.runBlockingTest {
+            val downstream = Channel<Dispatch.Value<String>>(Channel.UNLIMITED)
+            manager += downstream
+
+            val collection = async {
+                try {
+                    downstream.consumeAsFlow()
+                        .onEach { it.markDelivered() }
+                        .take(2)
+                        .toList()
+                        .map { it.value }
+                } finally {
+                    manager -= downstream
```

Comment:
i strongly dislike plusAssign / minusAssign overloads for this case.
makes it a lot harder to reason about in my opinon.